### PR TITLE
Conformer onnx fix

### PR DESCRIPTION
**Pull request by borisfom**

### CODE REVIEW

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 1]
```diff
@@ -163,7 +163,11 @@ def forward(self, x, pad_mask=None):
             x.masked_fill_(pad_mask.unsqueeze(1), 0.0)
 
         x = self.depthwise_conv(x)
-        x = self.batch_norm(x)
+        if torch.onnx.is_in_onnx_export():
+            with torch.cuda.amp.autocast(enabled=False):
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:167]: I thought BN was the primary focus of autocast. Does it fail for some reason under onnx context?

File: nemo/core/classes/exportable.py [hunk 2]
```diff
@@ -134,7 +134,7 @@ def export(
         dynamic_axes=None,
         check_tolerance=0.01,
     ):
-        my_args = locals()
+        my_args = locals().copy()
```
Comment (titu1994) [inline nemo/core/classes/exportable.py:137]: Shouldn't this be a deep copy ?

File: nemo/core/classes/exportable.py [hunk 3]
```diff
@@ -254,6 +253,7 @@ def _verify_runtime(
             import onnxruntime
         except (ImportError, ModuleNotFoundError):
             logging.warning(f"ONNX generated at {output}, not verified - please install onnxruntime.\n")
+            onnx.checker.check_model(onnx_model, full_check=True)
```
Comment (titu1994) [inline nemo/core/classes/exportable.py:256]: This will fail if import fails

**Review by titu1994 — COMMENTED** [review]
Some minor comments.

File: nemo/collections/asr/parts/submodules/conformer_modules.py (hunk 1 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/conformer_modules.py]: I am putting it under autocast(False) because otherwise it generates bad ONNX code

File: nemo/core/classes/exportable.py (hunk 2 above)
Comment (borisfom) [reply nemo/core/classes/exportable.py:65]: No its just to shut top warnings about locals list itself

File: nemo/core/classes/exportable.py (hunk 3 above)
Comment (borisfom) [reply nemo/core/classes/exportable.py]: onnxruntime and onyx are different packages

File: nemo/core/classes/exportable.py (hunk 2 above)
Comment (titu1994) [reply nemo/core/classes/exportable.py:65]: Ah ok

File: nemo/collections/asr/modules/conv_asr.py [hunk 4]
```diff
@@ -554,7 +555,7 @@ def __init__(
                 )
             )
             self.decoder_layers.append(nn.Conv1d(self.feat_hidden, self.feat_hidden, kernel_size=1, bias=True))
-            self.decoder_layers.append(nn.BatchNorm1d(self.feat_hidden, eps=1e-3, momentum=0.1))
+            self.decoder_layers.append(BatchNorm1dNoAutoCast(self.feat_hidden, eps=1e-3, momentum=0.1))
```
Comment (titu1994) [inline nemo/collections/asr/modules/conv_asr.py:558]: Won't this affect pretrained checkpoints ?

File: nemo/collections/asr/parts/submodules/multi_head_attention.py [hunk 5]
```diff
@@ -157,9 +158,8 @@ def rel_shift(self, x):
             x (torch.Tensor): (batch, nheads, time, 2*time-1)
         """
         b, h, qlen, pos_len = x.size()  # (b, h, t1, t2)
-
-        # need to add a column of zeros on the left side of last two dimensions to perform the relative shifting
-        x = torch.nn.functional.pad(x, pad=(1, 0, 0, 0))  # (b, h, t1, t2+1)
+        # need to add a column of zeros on the left side of last dimension to perform the relative shifting
+        x = torch.nn.functional.pad(x, pad=(1, 0))  # (b, h, t1, t2+1)
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/multi_head_attention.py:162]: @VahidooX  please check this formula

File: nemo/core/classes/exportable.py [hunk 6]
```diff
@@ -217,26 +145,9 @@ def export(
                         verbose=verbose,
                         export_params=export_params,
                         do_constant_folding=do_constant_folding,
-                        keep_initializers_as_inputs=keep_initializers_as_inputs,
                         dynamic_axes=dynamic_axes,
                         opset_version=onnx_opset_version,
                     )
-
-                    # Verify the model can be read, and is valid
-                    onnx_model = onnx.load(output)
```
Comment (titu1994) [inline nemo/core/classes/exportable.py:226]: No more verification ?

File: nemo/core/classes/exportable.py [hunk 7]
```diff
@@ -217,26 +145,9 @@ def export(
...
-                    onnx.checker.check_model(onnx_model, full_check=True)
-
-                    if check_trace:
```
Comment (titu1994) [inline nemo/core/classes/exportable.py:229]: Or trace check ?

File: nemo/core/classes/exportable.py [hunk 8]
```diff
@@ -285,88 +172,23 @@ def supported_export_formats(self):
...
-        replace_for_export(self, replace_1D_2D)
+        replace_for_export(self)
 
     def _export_teardown(self):
         """
         Override this method for any teardown code after export.
         """
         pass
 
-    def _wrap_forward_method(self):
```
Comment (titu1994) [inline nemo/core/classes/exportable.py:329]: Without these methods RNNT export will fail. Also why such a heavy refactor back to monolithic design of export? It will make future models harder to deal with

File: nemo/utils/neural_type_utils.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
+from collections import defaultdict
```
Comment (titu1994) [inline nemo/utils/neural_type_utils.py:1]: This can go into the Nemo/ core / utils directory instead of this one. Well it's just semantics it's fine here too but more appropriate there.

File: scripts/export.py [hunk 10]
```diff
@@ -141,6 +150,11 @@ def nemo_export(argv):
 
     logging.info("Successfully exported to {}".format(out))
 
+    del model
+
+    if args.runtime_check:
+        verify_runtime(out, input_list, input_dict, input_names, output_names, output_example)
+
```
Comment (titu1994) [inline scripts/export.py:157]: This should really be an implicit part of model export, not an external function. Most people will forget to verify their export in the code when they do model.export().

**Review by titu1994 — CHANGES_REQUESTED** [review]
This refactor breaks RNNT export totally. Even if that's fixed, the refactor to remove testing and runtime trace check as a part of export is necessary.

File: nemo/collections/asr/modules/conv_asr.py (hunk 4 above)
Comment (borisfom) [reply nemo/collections/asr/modules/conv_asr.py:558]: No they are fine - submodule has same name and parameters.

File: scripts/export.py (hunk 10 above)
Comment (borisfom) [reply scripts/export.py:157]: Well I agree in principle, but it should be available externally to be able to check models that won't allow 2 copies on GPU at the same time.

**Comment by borisfom** [comment]
Nothing is broken - my previous RNNT refactoring removed all those overridden methods. So this is cleanup part to remove unused hooks and keep code more readable.

File: nemo/collections/asr/parts/submodules/multi_head_attention.py (hunk 5 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/multi_head_attention.py:162]: We discussed that yesterday on slack

Comment (borisfom) [reply nemo/collections/asr/parts/submodules/multi_head_attention.py:162]: (1,0,0,0) is equivalent to (1,0) - just not misleading. Intent is to pad a single dimension

**Review by titu1994 — DISMISSED** [review]
RNNT seems fine for now.

File: nemo/utils/neural_type_utils.py (hunk 9 above)
Comment (borisfom) [reply nemo/utils/neural_type_utils.py]: done.

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 11]
```diff
@@ -262,7 +264,11 @@ def forward_for_export(self, audio_signal, length=None):
         if self.att_context_size[1] >= 0:
             att_mask = att_mask.tril(diagonal=self.att_context_size[1])
         att_mask = ~att_mask
-        pad_mask = ~pad_mask
+
+        if self.use_pad_mask:
+            pad_mask = ~pad_mask
+        else:
+            pad_mask = None
```
Comment (titu1994) [inline nemo/collections/asr/modules/conformer_encoder.py:271]: Just a note, None mode isn't used anywhere, is that ok ?

File: tests/collections/asr/test_asr_exportables.py [hunk 12]
```diff
@@ -119,13 +119,12 @@ def test_EncDecCitrinetModel_limited_SE_export_to_onnx(self, citrinet_model):
...
     def test_EncDecRNNTModel_export_to_onnx(self, citrinet_rnnt_model):
-        citrinet_rnnt_model.freeze()
-        model = citrinet_rnnt_model.train()
+        model = citrinet_rnnt_model.cuda()
 
         with tempfile.TemporaryDirectory() as tmpdir:
             fn = 'citri_rnnt.onnx'
             filename = os.path.join(tmpdir, fn)
-            model.export(output=filename, verbose=False, check_trace=True)
+            model.export(output=filename, verbose=False)
```
Comment (titu1994) [inline tests/collections/asr/test_asr_exportables.py:127]: Just wondering, why aren't we checking the trace is correct similar to before ?

**Review by titu1994 — APPROVED** [review]
LGTM, minor comments

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 11 above)
Comment (VahidooX) [reply nemo/collections/asr/modules/conformer_encoder.py:271]: What is the purpose of this use_pad_mask? It looks to be always True, right?

File: nemo/utils/export_utils.py [hunk 13]
```diff
@@ -20,6 +22,134 @@
...
+
+
+_EXT_DICT = {
+    ".pt": ExportFormat.TORCHSCRIPT,
+    ".ts": ExportFormat.TORCHSCRIPT,
+    ".onnx": ExportFormat.ONNX,
+}
+
+
+class BatchNorm1dNoAutoCast(nn.BatchNorm1d):
```
Comment (VahidooX) [inline nemo/utils/export_utils.py:40]: How about giving a more general/flexible name for this batch norm as it may get updated and changed over the time. For instance how about naming it something like BatchNorm1dExportable?

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 14]
```diff
@@ -95,33 +95,31 @@ def forward(self, x, att_mask=None, pos_emb=None, pad_mask=None):
         Returns:
             x (torch.Tensor): (B, T, d_model)
         """
+        dtype = x.dtype
         residual = x
         x = self.norm_feed_forward1(x)
         x = self.feed_forward1(x)
-        x = self.fc_factor * self.dropout(x) + residual
+        residual = residual + self.dropout(x) * self.fc_factor
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:102]: Was there any specific reason naming of x and residual are changed here?

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 15]
```diff
@@ -95,33 +95,31 @@ def forward(self, x, att_mask=None, pos_emb=None, pad_mask=None):
...
-        x = self.norm_feed_forward2(x)
+        x = self.norm_feed_forward2(residual)
         x = self.feed_forward2(x)
-        x = self.fc_factor * self.dropout(x) + residual
+        residual = residual + self.dropout(x) * self.fc_factor
 
-        x = self.norm_out(x)
-        return x
+        x = self.norm_out(residual)
+        return x.to(dtype=dtype)
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:122]: Can x here have a different type other than the type of the input?

File: nemo/utils/export_utils.py (hunk 13 above)
Comment (VahidooX) [reply nemo/utils/export_utils.py:40]: Many other models in nemo are using BatchNorm1d. Do we know if they have encountered any issues or not or why it just happened for Conformer and not the other models?

File: nemo/collections/asr/parts/submodules/conformer_modules.py [hunk 16]
```diff
@@ -160,24 +158,31 @@ def __init__(self, d_model, kernel_size, norm_type='batch_norm'):
...
+    def forward(self, x, pad_mask=None):
+        x = x.transpose(1, 2)
+        x = self.pointwise_conv1(x)
+        x = nn.functional.glu(x, dim=1)
+
+        if pad_mask is not None:
+            x.masked_fill_(pad_mask.unsqueeze(1), 0.0)
+
+            if torch.onnx.is_in_onnx_export():
+                with torch.cuda.amp.autocast(enabled=False):
```
Comment (VahidooX) [inline nemo/collections/asr/parts/submodules/conformer_modules.py:180]: Is this part necessary as we already have these autocasts and converisons already in the BatchNorm1dNoAutoCast?
BatchNorm1dNoAutoCast just misses the last .to(dtype=x.dtype) which can be added.

**Review by VahidooX — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 11 above)
Comment (borisfom) [reply nemo/collections/asr/modules/conformer_encoder.py:271]: I have prepared 'mask disabled' mode - I think you have suggested on Slack it could be skipped for inference - but numeric results were rather different so I left it disabled for now.

File: tests/collections/asr/test_asr_exportables.py (hunk 12 above)
Comment (borisfom) [reply tests/collections/asr/test_asr_exportables.py:127]: onnxruntime had some intermittent CUDA issues on my box so I disabled it just in case to avoid breaking CI randomly.

File: nemo/utils/export_utils.py (hunk 13 above)
Comment (borisfom) [reply nemo/utils/export_utils.py:40]: I had to also use it for CitriNet, yes. 
Regarding name - I think this way it's more descriptive.

File: nemo/collections/asr/parts/submodules/conformer_modules.py (hunk 15 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/conformer_modules.py:122]: Yes - with autocast, result of norm_out is fp32 when residual is x. I traced that. This is because autocast (correctly) leaves those weights in fp32. But we want to cast back to original type on return.

File: nemo/collections/asr/parts/submodules/conformer_modules.py (hunk 14 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/conformer_modules.py:102]: With ONNX opset 15, I have encountered some errors related to aliasing. Those changes fixed that.

File: nemo/collections/asr/parts/submodules/conformer_modules.py (hunk 16 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/conformer_modules.py:180]: We are not using BatchNorm1dNoAutoCast here - I had to use that helper class in Citrinet because there batch conv's forward() was not called explicitly. Here it is explicit, and I wanted to keep it fused with previous conv, so I did not use helper class.

**Comment by borisfom** [comment]
> While it does use the new change with "x = self.do_conv_norm(x.to(dtype=torch.float)).to(dtype=x.dtype)" it still finds half and not float. Am I doing something wrong?
@naor2013: Try with updated version. Are you using  .half() on the model? You should use torch.cuda.amp.autocast().

File: nemo/collections/asr/parts/submodules/conformer_modules.py (hunk 16 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/conformer_modules.py]: (changed to BatchNorm1dNoAutoCast for clarity)

**Comment by naor2013** [comment]
@borisfom Thank you! After your new commit, it worked!

**Comment by naor2013** [comment]
@borisfom Do you know why when I exported it without autocast, the onnx model had 3994 nodes but with autocast it had 4579? Are there more nodes to convert between the sizes or something?

**Comment by borisfom** [comment]
> @borisfom Do you know why when I exported it without autocast, the onnx model had 3994 nodes but with autocast it had 4579? Are there more nodes to convert between the sizes or something?

Yes. Many of those nodes will go away if you run 'polygraphy surgeon sanitize --fold-constants ...'

File: nemo/utils/export_utils.py [hunk 17]
```diff
@@ -219,7 +277,4 @@ def replace_for_export(model: nn.Module, replace_1D_2D: bool = False) -> nn.Modu
     Returns:
         model, possibly modified in-place
     """
-    replace_modules(model, default_Apex_replacements)
-    if replace_1D_2D:
-        # TODO: add squeeze/unsqueeze
-        replace_modules(model, default_1D_2D_replacements)
+    replace_modules(model, default_replacements)
```
Comment (titu1994) [inline nemo/utils/export_utils.py:280]: This will replace every instance of BN1d - even for citrinet, ContextNet, Jasper QN etc - will they be ok with this chance ?
Is this some regression in onnx itself for BN1d?

**Review by titu1994 — APPROVED** [review]
Overall it's good, but new change will now replace conv 1d in all older models - is that fine ?

**Review by VahidooX — APPROVED** [review]

**Comment by naor2013** [comment]
> > @borisfom Do you know why when I exported it without autocast, the onnx model had 3994 nodes but with autocast it had 4579? Are there more nodes to convert between the sizes or something?
> 
> Yes. Many of those nodes will go away if you run 'polygraphy surgeon sanitize --fold-constants ...'

@borisfom after running that it removed some nodes but still the autocast version had more nodes than the "regular" version. Why is that?

File: nemo/utils/export_utils.py [hunk 18]
```diff
@@ -20,6 +22,134 @@
...
+        import onnxruntime
+    except (ImportError, ModuleNotFoundError):
+        logging.warning(f"ONNX generated at {output}, not verified - please install onnxruntime.\n")
+        onnx.checker.check_model(onnx_model, full_check=True)
+        return
+
+    onnx_session_opt = onnxruntime.SessionOptions()
+    onnx_session_opt.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_ENABLE_ALL
+
+    providers = [
```
[the remark this answers is not in the document]
Comment (borisfom) [reply nemo/utils/export_utils.py:123]: Why, they are.

File: nemo/utils/export_utils.py (hunk 17 above)
Comment (borisfom) [reply nemo/utils/export_utils.py:280]: Yes they will be OK and that's what we want. BN1d weights are always in fp32, so those casts are no-ops, really - just helping to generate proper ONNX code.

File: nemo/collections/asr/parts/submodules/conformer_modules.py (hunk 15 above)
Comment (borisfom) [reply nemo/collections/asr/parts/submodules/conformer_modules.py:122]: Yes, that's what I see if I print out the type. not happening after calling other layers - maybe autocast treated the return from forward() differently.
